source/lac/trilinos_solver.cc

```diff
@@ -282,6 +291,143 @@ namespace TrilinosWrappers
   }
 
 
+  namespace internal
+  {
+    namespace
+    {
+      double
+      compute_residual (const Epetra_MultiVector *const residual_vector)
+      {
+        Assert(residual_vector->NumVectors() == 1,
+               ExcMessage("Residual multivector holds more than one vector"));
+        std_cxx11::array<double, 1> res_l2_norm;
+        residual_vector->Norm2(res_l2_norm.data());
+        return res_l2_norm[0];
```

Comment:
I think an `std::array` is a bit of an overkill here. We typically use a single variable `double res_l2_norm` and then pass in `&res_l2_norm` to `Norm2`. Have a look at the use of `Norm2` in `include/deal.II/lac/trilinos_vector.h`.